File: scripting-languages-winter-2017/perl/word-count.pl
#!/usr/bin/perl
# Witold Bobrowski ; Group 2
#
# Print word count for input files

use strict;
use warnings;
use bytes;
use Utils;

# CONSTANTS

my $FILE  = "file";
my $WORDS = "words";
my $LINES = "lines";
my $CHARS = "chars";
my $BYTES = "bytes";
my $INTS  = "ints";
my $NUMS  = "nums";

# VARIABLES

my $should_ignore_comments  = 0;
my $should_print_lines      = 0;
my $should_print_words      = 0;
my $should_print_characters = 0;
my $should_print_bytes      = 0;
my $should_print_ints       = 0;
my $should_print_nums       = 0;

my $total_lines = 0;
my $total_words = 0;
my $total_chars = 0;
my $total_bytes = 0;
my $total_ints  = 0;
my $total_nums  = 0;

my $longest_result = 0;

# SUBROUTINES

# Count lines, words, characters, bytes in files
sub word_count {
    my $lines_count = 0;
    my $words_count = 0;
    my $char_count  = 0;
    my $byte_count  = 0;
    my $int_count   = 0;
    my $num_count   = 0;
    open( my $data, '<:encoding(UTF-8)', $_[0] )
      or die "Could not open file '$_[0]' $!";
    while ( my $line = <$data> ) {

        if ( $should_ignore_comments && $line =~ /^\s*#/ ) {
            next;
        }
        my @words = split( " ", $line );
        $char_count += length($line);
        $byte_count += bytes::length($line);
        for my $word (@words) {
            $words_count += 1;
            if ( Utils::looks_like_number($word) ) {
                $num_count += 1;
                if ( Utils::looks_like_integer($word) ) {
                    $int_count += 1;
                }
            }
        }
        $lines_count += 1;
    }
    $total_lines += $lines_count;
    $total_words += $words_count;
    $total_chars += $char_count;
    $total_bytes += $byte_count;
    $total_ints  += $int_count;
    $total_nums  += $num_count;
    compare_to_longest_string(
        $lines_count, $words_count, $char_count,
        $byte_count,  $int_count,   $num_count
    );
    return (
        $FILE  => $_[0],
        $LINES => $lines_count,
        $WORDS => $words_count,
        $CHARS => $char_count,
        $BYTES => $byte_count,
        $INTS  => $int_count,
        $NUMS  => $num_count
    );
}

# Compare lenght of give string to the longest one recorded yet and if longer replace te result
sub compare_to_longest_string {
    for my $string (@_) {
        if ( length($string) > $longest_result ) {
            $longest_result = length($string);
        }
    }
}

# Evaluate all arguments passed to script, and turn on give flags if needed (if none is set, set default config)
sub evaluate_argumets {
    my @files;
    for my $i ( 0 .. $#ARGV ) {
        if ( "$ARGV[$i]" eq "-l" || "$ARGV[$i]" eq "--lines" ) {
            $should_print_lines = 1;
        }
        elsif ( "$ARGV[$i]" eq "-w" || "$ARGV[$i]" eq "--words" ) {
            $should_print_words = 1;
        }
        elsif ( "$ARGV[$i]" eq "-m" || "$ARGV[$i]" eq "--chars" ) {
            $should_print_characters = 1;
        }
        elsif ( "$ARGV[$i]" eq "-c" || "$ARGV[$i]" eq "--bytes" ) {
            $should_print_bytes = 1;
        }
        elsif ( "$ARGV[$i]" eq "-i" || "$ARGV[$i]" eq "--ints" ) {
            $should_print_ints = 1;
        }
        elsif ( "$ARGV[$i]" eq "-n" || "$ARGV[$i]" eq "--nums" ) {
            $should_print_nums = 1;
        }
        elsif ( "$ARGV[$i]" eq "-e" || "$ARGV[$i]" eq "--comments" ) {
            $should_ignore_comments = 1;
        }
        else {
            push @files, $ARGV[$i];
        }
    }
    if (   ( $should_print_lines eq 0 )
        && ( $should_print_words eq 0 )
        && ( $should_print_characters eq 0 )
        && ( $should_print_bytes eq 0 )
        && ( $should_print_ints eq 0 )
        && ( $should_print_nums eq 0 ) )
    {
        $should_print_lines      = 1;
        $should_print_words      = 1;
        $should_print_characters = 1;
    }
    return @files;
}

# Format result with appropriate amount of spaces
sub format_result {
    my $n      = $longest_result - length( $_[0] );
    my $spaces = ( ' ' x $n );
    return "$spaces$_[0]";
}

# Evaluate result from word_count subroutine to readable string
sub get_output {
    my %results = @_;
    my $output  = "";
    if ( $should_print_lines eq 1 ) {
        my $formated = format_result( $results{$LINES} );
        $output = "$output    $formated";
    }
    if ( $should_print_words eq 1 ) {
        my $formated = format_result( $results{$WORDS} );
        $output = "$output    $formated";
    }
    if ( $should_print_characters eq 1 ) {
        my $formated = format_result( $results{$CHARS} );
        $output = "$output    $formated";
    }
    if ( $should_print_bytes eq 1 ) {
        my $formated = format_result( $results{$BYTES} );
        $output = "$output    $formated";
    }
    if ( $should_print_ints eq 1 ) {
        my $formated = format_result( $results{$INTS} );
        $output = "$output    $formated";
    }
    if ( $should_print_nums eq 1 ) {
        my $formated = format_result( $results{$NUMS} );
        $output = "$output    $formated";
    }
    $output = "$output $results{$FILE}\n";
    return $output;
}

# Main program

my @files = evaluate_argumets();
for my $file (@files) {
    print get_output( word_count($file) );
}

if ( $#files > 0 ) {
    print get_output(
        (
            $FILE  => "total",
            $LINES => $total_lines,
            $WORDS => $total_words,
            $CHARS => $total_chars,
            $BYTES => $total_bytes,
            $INTS  => $total_ints,
            $NUMS  => $total_nums
        )
    );
}

File: scripting-languages-winter-2017/python/word-count.py
#!/usr/bin/env python3
# Witold Bobrowski ; Group 2
#
# Print word count for input files

import sys
import argparse as ap
import utils

# CONSTANTS
FILE, WORDS, LINES, CHARS, BYTES, INTS, NUMS = "file", "words", "lines", "chars", "bytes", "ints", "nums"

# VARIABLES
total_count = {LINES: 0, WORDS: 0, CHARS: 0, BYTES: 0, INTS: 0, NUMS: 0}
# Longest result
long_res = 0

# FUNCTIONS


def setup_parser():
    parser = ap.ArgumentParser()
    parser.add_argument('-e', '--comments', help="Will ignore lines starting with #",
                        dest='should_ignore_comments', action='store_true')
    parser.add_argument('-l', '--lines', help="Print number of lines",
                        dest='should_print_lines', action='store_true')
    parser.add_argument('-w', '--words', help="Print number of words",
                        dest='should_print_words', action='store_true')
    parser.add_argument('-m', '--chars', help="Print number of characters",
                        dest='should_print_characters', action='store_true')
    parser.add_argument('-c', '--bytes', help="Print number of bytes",
                        dest='should_print_bytes', action='store_true')
    parser.add_argument('-i', '--ints', help="Print number of integers",
                        dest='should_print_ints', action='store_true')
    parser.add_argument('-n', '--nums', help="Print number of numbers in any of the following formats: e/E/d/D/q/Q",
                        dest='should_print_nums', action='store_true')
    parser.add_argument('files', help="Files to evaluate.",
                        nargs='*', action="store")
    return parser.parse_args()


def evaluate_options(arguments):
    options = {arguments.should_print_lines, arguments.should_print_words, arguments.should_print_characters,
               arguments.should_print_bytes, arguments.should_print_ints, arguments.should_print_nums}
    for option in options:
        if option:
            return
    # If none of the options was chosen, turn on default ones
    arguments.should_print_lines = True
    arguments.should_print_words = True
    arguments.should_print_characters = True


def compare_to_longest_string(strings):
    global long_res
    for string in strings:
        long_res = len(str(string)) if len(
            str(string)) > long_res else long_res


def format_result(string):
    n = long_res - len(str(string))
    spaces = " " * n
    return spaces + str(string)


def word_count(file, should_ignore_comments):
    global total_count
    count = {FILE: file, LINES: 0, WORDS: 0, CHARS: 0, BYTES: 0, INTS: 0, NUMS: 0}
    with open(file, 'r') as file:
        count[FILE] = file.name
        for line in file:
            if should_ignore_comments and line.startswith("#"):
                continue
            words = line.split()
            count[LINES] += 1
            count[WORDS] += len(words)
            count[CHARS] += len(line)
            count[BYTES] += sys.getsizeof(line)
            for word in words:
                if utils.looks_like_number(word):
                    count[NUMS] += 1
                    if utils.looks_like_integer(word):
                        count[INTS] += 1
    for key in total_count:
        total_count[key] += count[key]
        compare_to_longest_string([total_count[key]])
    return count


def get_output(options, result):
    output = ""
    if options.should_print_lines:
        output += "    " + str(format_result(result[LINES]))
    if options.should_print_words:
        output += "    " + str(format_result(result[WORDS]))
    if options.should_print_characters:
        output += "    " + str(format_result(result[CHARS]))
    if options.should_print_bytes:
        output += "    " + str(format_result(result[BYTES]))
    if options.should_print_ints:
        output += "    " + str(format_result(result[INTS]))
    if options.should_print_nums:
        output += "    " + str(format_result(result[NUMS]))
    output += " " + result[FILE]
    return output


# Main
if __name__ == "__main__":
    arguments = setup_parser()
    evaluate_options(arguments)
    results = [word_count(file, arguments.should_ignore_comments)
               for file in arguments.files]
    for result in results:
        print(get_output(arguments, result))
    if len(results) > 1:
        total_count[FILE] = "total"
        print(get_output(arguments, total_count))
